server/forms/step.ts

```diff
@@ -0,0 +1,96 @@
+import { readFile } from 'fs/promises'
+import jsonLogic, { RulesLogic } from 'json-logic-js'
+import moment from 'moment'
+
+import { StepDefinition, ErrorMessages } from './interfaces'
+import Question from './question'
+
+const isDateString = (value: any): boolean => {
+  return moment(value, moment.ISO_8601).isValid()
+}
+
+const join = (values: Array<any>, seperator: string): string => {
+  return values.join(seperator)
+}
+
+jsonLogic.add_operation('isDateString', isDateString)
+jsonLogic.add_operation('join', join)
+
+export default class Step {
+  name: string = this.step.name
+
+  section: string = this.step.section
+
+  title: string = this.step.title
+
+  showTitle: boolean = this.step.showTitle
+
+  partial: string = this.step.partial
+
+  errorMessages: ErrorMessages
+
+  private constructor(private readonly step: StepDefinition, readonly body: any) {}
+
+  public static async initialize(name: string, body: any): Promise<Step> {
+    const json = await Step.readJson(name)
+
+    return new Step(json, body)
+  }
+
+  public nextStep(data: any): string {
+    return this.applyRule(this.step.nextStep, data)
+  }
+
+  public previousStep(data: any): string {
+    return this.applyRule(this.step.previousStep, data)
+  }
+
+  public valid(): boolean {
+    this.validate()
+
+    return Object.keys(this.errorMessages).length === 0
+  }
+
+  public questions(): Array<Question> {
+    return this.step.questions.map(question => new Question(this, question))
+  }
+
+  public allowedToAccess(data: any): boolean {
+    const rule = this.step.allowedToAccess
+
+    if (typeof rule === 'object') {
+      return jsonLogic.apply(rule, data)
+    }
+
+    return rule as boolean
```

Comment:
I don't think we need to cast this return type to a boolean, it could cause headaches down the line